guide_drafts/composing-diesel-applications.md

```diff
@@ -0,0 +1,246 @@
+Composing Applications with Diesel
+----------------------------------
+
+One of the main benefits of using a query builder over raw SQL
+is that you can pull bits of your query out into functions and reuse them.
+In this guide,
+we'll look at common patterns for extracting your code into re-usable pieces.
+We'll also look at best practices for how to structure your code.
+
+All of our code examples are based on code from crates.io,
+a real world application which uses Diesel extensively.
+All of our examples will be focused on functions which *return*
+queries or pieces of queries.
+None of these examples will include a function which takes a database
+connection.
+We will go into the benefits of this structure at the end of the guide.
+
+crates.io has a `canon_crate_name` SQL function
+which is always used when comparing crate names.
+Rather than continuously writing
+`canon_crate_name(crates::name).eq("some name")`,
+we can instead pull this out into a function.
+
+```
+use diesel::dsl::Eq;
+use diesel::types::Text;
+
+sql_function!(canon_crate_name, CanonCrateName, (x: Text) -> Text);
+
+type WithName<'a> = Eq<canon_crate_name<crates::name>, canon_crate_name<&'a str>>;
+
+fn with_name(name: &str) -> WithName {
+    canon_crate_name(crates::name).eq(canon_crate_name(name))
+}
+```
+
+Now when we want to find a crate by name, we can write
+`crates::table.filter(with_name("foo"))` instead.
+If we want to accept types other than a string,
+we can make the method generic.
+
+```
+use diesel::dsl::Eq;
+use diesel::types::Text;
+
+sql_function!(canon_crate_name, CanonCrateName, (x: Text) -> Text);
+
+type WithName<T> = Eq<canon_crate_name<crates::name>, canon_crate_name<T>>;
+
+fn with_name<T>(name: T) -> WithName<T>
+where
+    T: AsExpression<Text>,
+{
+    canon_crate_name(crates::name).eq(canon_crate_name(name))
+}
+```
+
+It's up to you whether you make your functions generic,
+or only take a single type.
+We recommend only making these functions generic if it's actually needed,
+since it requires additional bounds in your `where` clause.
+The bounds you need might not be clear,
+unless you are familiar with Diesel's lower levels.
+
+In these examples,
+we are using helper types from `diesel::dsl`
+to write the return type explicitly.
+Nearly every method in Diesel has a helper type like this.
+The first type parameter is the method receiver
+(the thing before the `.`).
+The remaining type parameters are the arguments to the method.
+If we want to avoid writing this return type,
+or dynamically return a different expression,
+we can box the value instead.
+
+```
+use diesel::pg::Pg;
+use diesel::types::Text;
+
+sql_function!(canon_crate_name, CanonCrateName, (x: Text) -> Text);
+
+fn with_name<'a, T>(name: T) -> Box<BoxableExpression<crates::table, Pg, SqlType = Bool> + 'a>
+where
+    T: AsExpression<Text>,
+    T::Expression: BoxableExpression<crates::table, Pg>,
+{
+    canon_crate_name(crates::name).eq(canon_crate_name(name))
```

Comment:
I think this is missing a `Box::new()` or something.

Reply:
Good catch

Reply:
Good catch